File: graphs/1192-critical-connections-in-a-network.py
from typing import List
import unittest

# https://leetcode.com/problems/critical-connections-in-a-network/

# python3 -m unittest graphs/1192-critical-connections-in-a-network.py


class Solution(unittest.TestCase):
    # Approach: Tarjan's Algorithm
    # Time: O(V + E), Space: O(V + E)
    def criticalConnections(self, n: int, connections: List[List[int]]) -> List[List[int]]:
        graph: List[List[int]] = [[] for _ in range(n)]
        for u, v in connections:
            graph[u].append(v)
            graph[v].append(u)

        dfn = [-1] * n
        low = [-1] * n
        bridges: List[List[int]] = []
        timer = 0

        def dfs(node: int, parent: int):
            nonlocal timer, bridges

            dfn[node] = timer
            low[node] = timer
            timer += 1

            for neighbor in graph[node]:
                if neighbor == parent:
                    continue

                if dfn[neighbor] == -1:
                    dfs(neighbor, node)
                    low[node] = min(low[node], low[neighbor])

                    if low[neighbor] > dfn[node]:
                        bridges.append([node, neighbor])
                else:
                    low[node] = min(low[node], dfn[neighbor])

        dfs(0, -1)
        return bridges

    def test(self):
        for n, connections, expected in [
            (4, [[0, 1], [1, 2], [2, 0], [1, 3]], [[1, 3]]),
            (2, [[0, 1]], [[0, 1]]),
        ]:
            output = self.criticalConnections(n, connections)
            output.sort()
            expected.sort()
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: graphs/1192-critical-connections-in-a-network.go
package graphs

// https://leetcode.com/problems/critical-connections-in-a-network/

// Approach: Tarjan's Algorithm for finding bridges
// Use DFS with discovery time and low link values
// Time: O(V + E)
// Space: O(V + E)
func criticalConnections(n int, connections [][]int) [][]int {
	// Build adjacency list
	graph := make([][]int, n)

	for _, conn := range connections {
		u, v := conn[0], conn[1]
		graph[u] = append(graph[u], v)
		graph[v] = append(graph[v], u)
	}

	dfn := make([]int, n)
	low := make([]int, n)

	for i := range dfn {
		dfn[i] = -1
	}

	bridges := [][]int{}
	timer := 0

	var dfs func(int, int)

	dfs = func(node, parent int) {
		dfn[node] = timer
		low[node] = timer
		timer++

		for _, neighbor := range graph[node] {
			if neighbor == parent {
				continue
			}

			if dfn[neighbor] == -1 {
				dfs(neighbor, node)
				low[node] = min(low[node], low[neighbor])

				// Bridge found
				if low[neighbor] > dfn[node] {
					bridges = append(bridges, []int{node, neighbor})
				}
			} else {
				low[node] = min(low[node], dfn[neighbor])
			}
		}
	}

	dfs(0, -1)

	return bridges
}
